core/trie/proof.go

```diff
@@ -54,623 +55,586 @@ func (e *Edge) Len() uint8 {
-		nodeLeft, leftExist := nodeHashes[*currNode.LeftHash]
-		nodeRight, rightExist := nodeHashes[*currNode.RightHash]
+		switch node := proofNode.(type) {
+		case *Binary: // Binary nodes represent left/right choices
+			if key.Len() <= curPos {
+				return nil, fmt.Errorf("key length less than current position, key length: %d, current position: %d", key.Len(), curPos)
+			}
+			// Determine the next node to traverse based on the next bit position
+			expectedHash = node.LeftHash
+			if key.IsBitSet(keyLen - curPos - 1) {
+				expectedHash = node.RightHash
+			}
+			curPos++
+		case *Edge: // Edge nodes represent paths between binary nodes
+			if !verifyEdgePath(&key, node.Path, curPos) {
+				return &felt.Zero, nil
+			}
 
-		if leftExist && rightExist {
-			return
-		} else if leftExist {
-			traverseNodes(nodeLeft, path, nodeHashes)
-		} else if rightExist {
-			traverseNodes(nodeRight, path, nodeHashes)
+			// Move to the immediate child node
+			curPos += node.Path.Len()
+			expectedHash = node.Child
 		}
-	case *Edge:
-		edgeNode, exist := nodeHashes[*currNode.Child]
-		if exist {
-			traverseNodes(edgeNode, path, nodeHashes)
+
+		// We've consumed all bits in our path
+		if curPos >= keyLen {
+			return expectedHash, nil
 		}
 	}
 }
 
-// MergeProofPaths removes duplicates and merges proof paths into a single path
-// merges paths in the specified order [commonNodes..., leftNodes..., rightNodes...]
-// ordering of the merged path is not important
-// since SplitProofPath can discover the left and right paths using the merged path and the rootHash
-func MergeProofPaths(leftPath, rightPath []ProofNode, hash hashFunc) ([]ProofNode, *felt.Felt, error) {
-	merged := []ProofNode{}
-	minLen := min(len(leftPath), len(rightPath))
-
-	if len(leftPath) == 0 || len(rightPath) == 0 {
-		return merged, nil, errors.New("empty proof paths")
-	}
-
-	if !leftPath[0].Hash(hash).Equal(rightPath[0].Hash(hash)) {
-		return merged, nil, errors.New("roots of the proof paths are different")
+// VerifyRangeProof checks whether the given key-value pairs and range proof are valid for the given root hash.
+// The range of key-value pairs should be consecutive (no gap inside) and monotonic increasing.
+// The given range proof contains 2 edge proofs. One is for the first key and the other is for the last key.
+// Both of them can be existent or non-existent.
+// This function also handles the following special cases:
+//
+//   - All elements proof: In this case the proof can be nil, but the range should be all the leaves in the trie.
+//   - One element proof: In this case, both left and right edge proofs are the same and the range should contain only one element.
+//   - Zero element proof: In this case a single edge proof is enough to verify. If there are still more elements, the proof is invalid.
+//
+// If the range proof is invalid, this function returns false and an error.
+// Otherwise, it will return a flag to indicate if there are more elements.
+//
```

Comment:
```suggestion
// VerifyRangeProof checks the validity of given key-value pairs and range proof against a provided root hash.
// The key-value pairs should be consecutive (no gaps) and monotonically increasing.
// The range proof contains two edge proofs: one for the first key and another for the last key.
// Both edge proofs can be for existent or non-existent keys.
// This function handles the following special cases:
//
//   - All elements proof: The proof can be nil if the range includes all leaves in the trie.
//   - Single element proof: Both left and right edge proofs are identical, and the range contains only one element.
//   - Zero element proof: A single edge proof suffices for verification. The proof is invalid if there are additional elements.
//
// The function returns a boolean indicating if there are more elements and an error if the range proof is invalid.
//
```
Please double-check whether I am not going too far here.